### PR TITLE
Fixes #1388 by correctly checking whether we are in ec2

### LINKED ISSUE #1388: Make arctic use ec2 role credentials when connecting to s3
**Is your feature request related to a problem? Please describe.**
Connecting to an s3 storage from an ec2 which has access to the s3 storage via an ec2 role doesn't work without explicitly passing credentials.

**Describe the solution you'd like**
It would be nice if `aws_auth=True` automatically checks for ec2 role permissions for s3 and uses those if available.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by IvoDD**
- The existing tcp_ping check didn't work.
- Uses the suggested by aws docs to query aws imds using libcurl.
- Tested on an ec2 machine manually to verify it works.

#### Reference Issues/PRs
Fixes #1388 

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [ ] Have you updated the relevant docstrings, documentation and copyright notice?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [ ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: cpp/arcticdb/storage/s3/ec2_utils.hpp
Comment (vasil-pashov) [inline cpp/arcticdb/storage/s3/ec2_utils.hpp:1]: Can we move function implementations to a .cpp file?

File: cpp/arcticdb/storage/s3/ec2_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        curl_slist *headers = nullptr;
+        headers = curl_slist_append(headers, "X-aws-ec2-metadata-token-ttl-seconds: 21600");
+        curl_easy_setopt(curl, CURLOPT_URL, "http://169.254.169.254/latest/api/token");
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+        curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PUT");
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
+
+        // We use a timeout of 100ms because it should be enough for IMDS to respond. Be wary of increasing the timeout
+        // since we need to wait the whole 100ms every time we init the aws sdk outside of aws ec2.
+        curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, 100);
```
Comment (poodlewars) [inline cpp/arcticdb/storage/s3/ec2_utils.hpp:35]: Make timeout configurable with the normal `ConfigsMap` approach? Useful to have a "get out of jail card" if we realise it's set too low and we mistakenly claim we are not on EC2, for example

File: cpp/arcticdb/storage/s3/ec2_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        if (res == CURLE_OK){
+            return true;
+        }
+        return false;
+    }
+    return false;
+}
+
+bool is_running_inside_aws_fast(){
```
Comment (poodlewars) [inline cpp/arcticdb/storage/s3/ec2_utils.hpp:63]: Worth some debug logging for what we decide? There is some in the caller for when we decide we are _not_ running in AWS, but not for when we _are_

File: cpp/arcticdb/storage/s3/ec2_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+// A fast check to identify whether we're running on AWS EC2.
+// According to AWS docs a reliable way to identify if we're in EC2 is to query the instance metadata service:
+// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html
+//
+// Since there are two versions IMDSv1 and IMDSv2 we first try to connect to v2 and if we fail then attempt the legacy
+// v1 connection. If both fail we're most likely running outside of EC2 (unless IMDS is under heavy load and takes more
+// than 100ms to respond)
+bool has_connection_to_ec2_imds(){
+    CURL *curl = curl_easy_init();
```
Comment (poodlewars) [inline cpp/arcticdb/storage/s3/ec2_utils.hpp:19]: Might be more readable to shortcircuit, `if (!curl) // return false immediately` and then the main implementation doesn't need to be nested.

**Comment by poodlewars** [comment]
Have you tested how this behaves when _not_ running in EC2?

It would be nice to have some automated tests for this, but it is a difficult thing to test. If we make the IP address configurable (for testing) it should be possible to spin up a mock server (in a pytest) and hit that with your curl requests. Then the hard bit is how to assert that we've set up the AWS SDK properly. Could you experiment a bit and see if we can get this testable in a reasonable way, and if it's a nightmare then I can live with just relying on manual testing.

File: python/tests/integration/arcticdb/test_s3.py [hunk 4]
```diff
@@ -24,3 +26,25 @@ def test_s3_storage_failures(mock_s3_store_with_error_simulation):
 
     with pytest.raises(StorageException, match="Unexpected error: S3Error#17"):
         lib.read(symbol_fail_read)
+
+
+# TODO: To make this test run alongside other tests we'll need to:
```
Comment (poodlewars) [inline python/tests/integration/arcticdb/test_s3.py:31]: Could something like this help? https://github.com/man-group/ArcticDB/pull/1409/files Could recreate it in a fixture and put it back to its normal state after the test finishes

**Review by poodlewars — APPROVED** [review]

File: python/tests/integration/arcticdb/test_s3.py [hunk 5]
```diff
@@ -24,3 +26,25 @@ def test_s3_storage_failures(mock_s3_store_with_error_simulation):
...
+# - Figure out how to do AWS::InitAPI multiple times in the same process. Currently we use std::call_once to ensure we
+#   we do this exactly once.
+# - Perform cleanup after tests (unset AWS_EC2_METADATA_DISABLED after done with each test, unset the runtime config
+#   "EC2.TestIMDSEndpointOverride" to make follow up tests work as expected
+@pytest.mark.skip(reason="Test only works if not run along other tests.")
+@pytest.mark.parametrize("run_on_aws", [True, False])
+def test_s3_running_on_aws_fast_check(lib_name, s3_storage_factory, run_on_aws):
```
Comment (poodlewars) [inline python/tests/integration/arcticdb/test_s3.py:38]: Nice job getting this working!

**Pull request merged**